new recipes

jcubic opened the pull request:
Few recipes (some from my implementation as-is and some modified). The best IMHO is `sorted?` it can be renamed to `ordered?` using two other recipes.

I think that my `take` function, which is part of SRFI-1, is better because the list should be the last argument, otherwise you can't `curry` easily that function. The same argument I've seen in a video some time ago about Lodash (and underscore) JavaScript libraries, that had broken arguments and you could not use `curry` with them. Because an array was the first argument.

jcubic committed:
new recipes

jcubic commented:
I have in fact implementation this function in my Scheme, I took those functional programming tools, from RambdaJS library that have very well defined API for function manipulation.
```scheme
(define (flip fn)
  "(flip fn)

   Higher order function that return new function where first two arguments are swapped.

   Example:

     (define first (curry (flip vector-ref) 0))
     (first #(1 2 3))
     ;; ==> 1"
  (typecheck "flip" fn "function")
  (lambda (a b . rest)
    (apply fn b a rest)))
```

lassik commented:
[SRFI 197: Pipeline Operators](https://srfi.schemers.org/srfi-197/srfi-197.html) solves the problem another way: you type an underscore `_` in the place where the argument should be inserted.

lassik commented:
The argument order of `take` is indeed cumbersome. `map` and `for-each` are even worse! But defining new standard procedures with the order flipped would probably cause more confusion than it solves, since we would end up with two procedures that do the same job. Using `flip` or `chain` in one's own program is a good solution.

jcubic commented:
Will change take to be the same as inn SRFI then.

jcubic committed:
update recipes

* use take function argument order from SRFI-1
* use loop name in let
* add another example to seq-map
* use another solution for map-over-sequence of elements recipe

jcubic commented on recipes/select-first-n-elements-from-list.md:
````diff
@@ -0,0 +1,28 @@
+
+```schem
+(define (take n lst)
+  (let iter ((result '()) (i n) (lst lst))
````
fixed, also change the name `iter` to `loop`

jcubic commented on recipes/map-over-sequance-of-elements-inside-list.md:
````diff
@@ -0,0 +1,39 @@
+```scheme
+(seq-map 2 < '(1 2 3 4))
+;; ==> (#t #t #t)
+```
````
added another example for N=3.

jcubic commented:
@lassik I also added another solution for the map recipe since you can understand the problem in two ways.

jcubic committed:
update example for map-over-sequence recipe

lassik commented on recipes/select-first-n-elements-from-list.md:
````diff
@@ -0,0 +1,26 @@
+```schem
+(define (take lst n)
+  (let loop ((result '()) (i n) (lst lst))
+    (if (or (null? lst) (<= i 0))
````
[SRFI 1 `take`](https://srfi.schemers.org/srfi-1/srfi-1.html#take) doesn't say what happens when the list has less than `n` elements. [The sample implementation](https://github.com/scheme-requests-for-implementation/srfi-1/blob/41993e344c1f5557cc125e9696cd746d2a1e46ac/srfi-1-reference.scm#L482) implicitly raises an exception since `cdr` will fail on a non-list. I also tried `(import (srfi 1))` in three Scheme implementations, and they all raise an exception. Hence, our `take` should probably do so too.

jcubic replied:
I can make two variations if you want. Because I think that is more useful:

```scheme
(take '(1 2) 10)
;; ==> (1 2)
```

lassik commented on recipes/map-over-sequance-of-elements-inside-list.md:
````diff
@@ -0,0 +1,73 @@
+```scheme
+(define (seq-map n fn seq-list)
+  (let loop ((seq-list seq-list) (result '()))
+    (if (null? seq-list)
````
Is it intentional that `seq-map` will call `fn` with lists shorter than `n`?

```Scheme
(define (take lst n)
  (let loop ((result '()) (i n) (lst lst))
    (if (or (null? lst) (<= i 0))
        (reverse result)
        (loop (cons (car lst) result) (- i 1) (cdr lst)))))

(define (seq-map n fn seq-list)
  (let loop ((seq-list seq-list) (result '()))
    (if (null? seq-list)
        (reverse result)
        (let ((next-list (take seq-list n)))
          (loop (cdr seq-list) (cons (apply fn next-list) result))))))

(define (seq-map/alternative n fn seq-list)
  (define (maybe-take list n)
    (let loop ((n n) (list list) (result '()))
      (cond ((<= n 0) (reverse result))
            ((not (pair? list)) #f)
            (else (loop (- n 1) (cdr list) (cons (car list) result))))))
  (let loop ((seq-list seq-list) (result '()))
    (let ((next-list (maybe-take seq-list n)))
      (if (not next-list)
          (reverse result)
          (loop (cdr seq-list) (cons (apply fn next-list) result))))))

(define (disp . args)
  (for-each display args)
  (newline))

(let ((example '(1 2 3 4 5)))
  (seq-map 3 disp example)
  (newline)
  (seq-map/alternative 3 disp example))
```

displays:

```
123
234
345
45
5

123
234
345
```

jcubic replied:
Yes, it's intentional. I can add an alternative version that do that. but the problem is that if you only check 3 elements you can have elements that will never get into the function.

jcubic replied:
What about if you have 2 elements and run it against 3 elements seq-map do you want to test the length of the list ot check if there are more then 3 elements? Or just call function on that two items?

jcubic replied:
I think that best would be to create two versions, the user will decide which one he want.

jcubic committed:
update the recipe

* change name to sublist-map
* add suggested example

jcubic committed:
update every? and take

jcubic committed:
change sorted? to ordered?

Make the name different than SRFI to not confuse people

jcubic commented:
Hi, sorry about the late response. I was busy with other things. I've updated the code. in some cases, I used two implementations, added a link to SRFI, or renamed the function.

jcubic committed:
update `Map over n consecutive elements` recipe

jcubic committed:
grammar

jcubic commented:
I've changed the `Map over n consecutive elements` recipe to only be called to full length sub lists.

What else needs to be done here?

jcubic committed:
two more recipes based on #43

arthurgleckler committed:
Merge pull request #71 from schemedoc/recipes-2

Recipes 2

lassik commented:
Sorry, I'm too slow to respond to these updates in a timely manner. Scheme.org should add content faster, and I should not be a gatekeeper that slows it down.

If you don't get a response from anyone within a day, just add what you think is needed. We can always change and revert things later.

We use rsync to upload everything to our Linux server. Let me know if you want an account there.

lassik commented:
Now up at https://cookbook.scheme.org/

jcubic commented:
Do you manually copy the files? Maybe it's good idea to setup GitHub action that will upload the files to the server.

Let me know if you need help with this. I can setup the workflow, but I need to know how to build the files, and you will need to add SSH key into secrets of the repo, I'm not sure if you can do this on org level. I use scp in one of the project to build website and upload to the server/hosting.

You can see similar workflow here (for my personal blog):

https://github.com/jcubic/jankiewicz/blob/master/.github/workflows/build.yaml

It requires Private SSH key as secret in the repo. But only the person that add the secret can see it, so it's safe for one admin to add the secret and not need to worry that other will see it.

arthurgleckler commented:
On Sat, Apr 6, 2024 at 10:05 AM Jakub T. Jankiewicz <
***@***.***> wrote:

> Do you manually copy the files? Maybe it's good idea to setup GitHub
> action that will upload the files to the server after merge.
>
> No, there's a script
<https://github.com/schemedoc/docs.scheme.org/blob/master/scripts/upload.sh>
that
does it.

> Let me know if you need help with this. I can setup the workflow, but I
> need to know how to build the files, and you will need to add SSH key into
> secrets of the repo, I'm not sure if you can do this on org level. I use
> scp in one of the project to build website and upload to the server/hosting.
>
> I'll leave it up to Lassi to decide whether to do this.  Our current
system doesn't require much work.  We just have to keep up with the
reviews.  I've been trying to keep up, but must have missed some.

> Message ID: ***@***.***>
>

lassik commented:
If we can make a Docker container that builds the HTML files (using our Scheme scripts) and uploads them using rsync, that would be ideal IMHO. We can make a dedicated SSH account for CI jobs if it helps.

lassik commented:
> We just have to keep up with the reviews. I've been trying to keep up, but must have missed some.

It's inevitable. Lisp doesn't attract enough people with the right personality to maintain good websites on a fast schedule.

The GNU sites are the only professional-looking Lisp sites. Even Clojure doesn't have a sleek website though many commercial companies are using it.

[Do-ocracy](https://communitywiki.org/wiki/DoOcracy) in all the details has to be the main driver of Scheme.org -- once we have written the charter to ensure that the basic structure and technical choices behind the site enjoy wide agreement.

arthurgleckler commented:
On Sat, Apr 6, 2024 at 12:30 PM lassik ***@***.***> wrote:

> We just have to keep up with the reviews. I've been trying to keep up, but
> must have missed some.
>
> It's inevitable. Lisp doesn't attract enough people with the right
> personality to maintain good websites on a fast schedule.
>
> Trying to push Jakub's new entry, I've realized that I haven't been
keeping up at all.  I was thinking of Schemedoc when I wrote the above.  I
can't even get the www.sh in the cookbook to run ("cannot import from
undefined module: colorize").

> Message ID: ***@***.***>
>

arthurgleckler commented:
On Sat, Apr 6, 2024 at 12:36 PM Arthur A. Gleckler ***@***.***>
wrote:


> Trying to push Jakub's new entry, I've realized that I haven't been
> keeping up at all.  I was thinking of Schemedoc when I wrote the above.  I
> can't even get the www.sh in the cookbook to run ("cannot import from
> undefined module: colorize").
>

Okay, it's up.  I've updated the comments to explain what to do.

> Message ID: ***@***.***>
>>
>

lassik commented:
Sorry, the outdated installation instructions are my fault. Thank you for fixing them.

jcubic commented:
I will create an issue for the workflow, so it's easier to find.
